From a31561c392adf5abcda0284e8049fafcb3701585 Mon Sep 17 00:00:00 2001
From: Xaver Hugl <xaver.hugl@gmail.com>
Date: Mon, 22 Apr 2024 13:20:09 +0200
Subject: [PATCH] backends/drm: use dumb buffers for the cursor on virtual
 machines

Apparently not all VM drivers handle dmabufs on the cursor plane correctly

BUG: 485827
(cherry picked from commit c14c61f74595c94d3533093ade45aea06ae45e28)
---
 src/backends/drm/drm_egl_cursor_layer.cpp | 2 +-
 src/backends/drm/drm_gpu.cpp              | 5 +++++
 src/backends/drm/drm_gpu.h                | 1 +
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/backends/drm/drm_egl_cursor_layer.cpp b/src/backends/drm/drm_egl_cursor_layer.cpp
index c20886e0b49..73b827e5782 100644
--- a/src/backends/drm/drm_egl_cursor_layer.cpp
+++ b/src/backends/drm/drm_egl_cursor_layer.cpp
@@ -42,7 +42,7 @@ static OutputTransform drmToOutputTransform(DrmPipeline *pipeline)
 
 EglGbmCursorLayer::EglGbmCursorLayer(EglGbmBackend *eglBackend, DrmPipeline *pipeline)
     : DrmPipelineLayer(pipeline)
-    , m_surface(pipeline->gpu(), eglBackend, pipeline->gpu()->atomicModeSetting() ? EglGbmLayerSurface::BufferTarget::Linear : EglGbmLayerSurface::BufferTarget::Dumb, EglGbmLayerSurface::FormatOption::RequireAlpha)
+    , m_surface(pipeline->gpu(), eglBackend, pipeline->gpu()->atomicModeSetting() && !pipeline->gpu()->isVirtualMachine() ? EglGbmLayerSurface::BufferTarget::Linear : EglGbmLayerSurface::BufferTarget::Dumb, EglGbmLayerSurface::FormatOption::RequireAlpha)
 {
 }
 
diff --git a/src/backends/drm/drm_gpu.cpp b/src/backends/drm/drm_gpu.cpp
index c5d258a2d7c..4f9803ea7c6 100644
--- a/src/backends/drm/drm_gpu.cpp
+++ b/src/backends/drm/drm_gpu.cpp
@@ -714,6 +714,11 @@ bool DrmGpu::isVmwgfx() const
     return m_isVmwgfx;
 }
 
+bool DrmGpu::isVirtualMachine() const
+{
+    return m_isVirtualMachine;
+}
+
 bool DrmGpu::isRemoved() const
 {
     return m_isRemoved;
diff --git a/src/backends/drm/drm_gpu.h b/src/backends/drm/drm_gpu.h
index 0cc6b2126a3..b5a42b32bad 100644
--- a/src/backends/drm/drm_gpu.h
+++ b/src/backends/drm/drm_gpu.h
@@ -80,6 +80,7 @@ public:
     bool isI915() const;
     bool isNVidia() const;
     bool isVmwgfx() const;
+    bool isVirtualMachine() const;
     gbm_device *gbmDevice() const;
     EglDisplay *eglDisplay() const;
     DrmBackend *platform() const;
-- 
GitLab

